apps/meteor/server/publications/messages.ts

```
@@ -7,65 +7,228 @@ import type { FindOptions } from 'mongodb';
-			throw new Meteor.Error('error-not-allowed', 'Not allowed', {
-				method: 'messages/get',
-			});
+			throw new Meteor.Error('error-not-allowed', 'Not allowed', { method: 'messages/get' });
 		}
 
-		const options: FindOptions<IMessage> = {
-			sort: {
-				ts: -1,
-			},
-		};
-
-		if (lastUpdate instanceof Date) {
-			return {
-				updated: await Messages.findForUpdates(rid, lastUpdate, {
-					sort: {
-						ts: -1,
-					},
-				}).toArray(),
-				deleted: await Messages.trashFindDeletedAfter(lastUpdate, { rid }, { ...options, projection: { _id: 1, _deletedAt: 1 } }).toArray(),
-			};
+		if (type && !['UPDATED', 'DELETED'].includes(type)) {
+			throw new Meteor.Error('error-type-param-not-supported', 'The "type" parameter must be either "UPDATED" or "DELETED"');
+		}
+
+		if ((next || previous) && !type) {
+			throw new Meteor.Error(
+				'error-type-param-required',
+				'The "type" parameter is required when using the "next" or "previous" parameters',
+			);
+		}
+
+		if (next && previous) {
+			throw new Meteor.Error('error-cursor-conflict', 'You cannot have both "next" and "previous" parameters');
 		}
 
-		return Meteor.callAsync('getChannelHistory', {
-			rid,
-			latest: latestDate,
-			oldest: oldestDate,
-			inclusive,
-			count,
-			unreads,
-		});
+		if ((next || previous) && lastUpdate) {
+			throw new Meteor.Error(
+				'error-cursor-and-lastUpdate-conflict',
+				'The attributes "next", "previous" and "lastUpdate" cannot be used together',
+			);
+		}
+
+		const hasCursorPagination = !!((next || previous) && count !== null && type);
+
+		if (!hasCursorPagination && !lastUpdate) {
+			return Meteor.callAsync('getChannelHistory', {
+				rid,
+				latest: latestDate,
+				oldest: oldestDate,
+				inclusive,
+				count,
+				unreads,
+			});
+		}
+
+		const response = lastUpdate
+			? await handleWithoutPagination(rid, lastUpdate)
+			: await handleCursorPagination(type ?? 'UPDATED', rid, count, next, previous);
```

I'd say it would be better to throw an error if type is undefined rather than assume it as 'UPDATED' as default.